conjure-java-jackson-serialization/src/main/java/com/palantir/conjure/java/serialization/CollectionValidationModule.java

```diff
@@ -0,0 +1,146 @@
+ * </ul>
+ */
+final class CollectionValidationModule extends SimpleModule {
+
+    private static final Logger log = LoggerFactory.getLogger(CollectionValidationModule.class);
+
+    CollectionValidationModule() {
+        super("collection-validation");
+        addAbstractTypeMapping(List.class, ValidatingArrayList.class);
+        addAbstractTypeMapping(Set.class, ValidatingHashSet.class);
+        addAbstractTypeMapping(Map.class, ValidatingHashMap.class);
+    }
+
+    /** Logs noisily if null values are received. */
+    static final class ValidatingArrayList<T> extends ArrayList<T> {
+
+        @Override
+        public boolean add(T value) {
+            return super.add(logIfNull(value));
+        }
+
+        @Override
+        public void add(int index, T value) {
+            super.add(index, logIfNull(value));
+        }
+
+        @Override
+        public boolean addAll(Collection<? extends T> collection) {
+            collection.forEach(CollectionValidationModule::logIfNull);
+            return super.addAll(collection);
+        }
+
+        @Override
+        public boolean addAll(int index, Collection<? extends T> collection) {
+            collection.forEach(CollectionValidationModule::logIfNull);
+            return super.addAll(index, collection);
+        }
+
+        @Override
+        public T set(int index, T value) {
+            return super.set(index, logIfNull(value));
+        }
+    }
+
+    /** Logs noisily if null values are received. */
+    static final class ValidatingHashSet<T> extends HashSet<T> {
+
+        @Override
+        public boolean add(T value) {
+            return super.add(logIfNull(value));
+        }
+
+        @Override
+        public boolean addAll(Collection<? extends T> collection) {
+            collection.forEach(CollectionValidationModule::logIfNull);
+            return super.addAll(collection);
+        }
+    }
+
+    /** Logs noisily if null values are received. */
+    static final class ValidatingHashMap<K, V> extends HashMap<K, V> {
+        private static final BiConsumer<Object, Object> validator = (key, value) -> logIfNull(value);
+
+        @Override
+        public V put(K key, V value) {
+            V previousValue = super.put(key, logIfNull(value));
+            if (previousValue != null) {
+                // Avoid the performance cost of Throwable.fillInStackTrace if WARN has been disabled
+                if (log.isWarnEnabled()) {
+                    // TODO(ckozak): Log key and values as unsafe arguments. This would require us to take a
+                    // dependency on safe-logging.
+                    log.warn("Detected duplicate map keys which are not allowed by Conjure",
+                            new IllegalArgumentException());
+                }
+                assert false : "Duplicate values for the same key are not allowed by Conjure. Key '"
+                        + key + "' values ['" + value + "', '" + previousValue + "']";
```

Comment:
performance nit: might want to pull the logic inside the if block out to a separate method so that the main `put` method bytecode is small and gets JITed & inlined, while the duplicate logging & assert is out-lined in separate slow path.

Reply:
Good idea.
I'm curious if you have any tricks for measuring whether a specific method hits the 35 byte sweet spot?

Reply:
`javap` for the classfile is quickest, but I also have been maintaining a [fork of the `jitwatch-intellij` plugin](https://github.com/schlosna/jitwatch-intellij) that you can run your unit tests/sample load.

Before (`put` was bytecode size 100) :
![image](https://user-images.githubusercontent.com/54594/54381043-2ea85080-4663-11e9-9545-550cef8d22cf.png)

After (`put` is byte code size 23 so can get inlined):
![image](https://user-images.githubusercontent.com/54594/54381241-aa0a0200-4663-11e9-8e61-a27289ced234.png)

I think we may want to extract out the `onNull` handling as well to optimize happy path:
```
    private static <T> T logIfNull(@Nullable T value) {
        // Log noisily in production
        if (value == null) {
            onNull();
        }
        return value;
    }

    private static void onNull() {
        // Tests should fail, but we're not ready to cause breaks in deployed systems yet.
        assert false : "Null values are not allowed by Conjure";

        // Avoid the performance cost of Throwable.fillInStackTrace if WARN has been disabled
        if (log.isWarnEnabled()) {
            log.warn("Detected a null value, null values are not allowed by Conjure", new IllegalArgumentException());
        }
    }
```

```
private static Object logIfNull(Object)
    0: aload_0         
    1: ifnonnull       7    
    4: invokestatic    #12  // Method onNull:()V
    7: aload_0         
    8: areturn         

private static void onNull()
     0: getstatic       #13  // Field $assertionsDisabled:Z
     3: ifne            16   
     6: new             #14  // class java/lang/AssertionError
     9: dup             
    10: ldc             #15  // String Null values are not allowed by Conjure
    12: invokespecial   #16  // Method java/lang/AssertionError."<init>":(Ljava/lang/Object;)V
    15: athrow          
    16: getstatic       #1   // Field log:Lorg/slf4j/Logger;
    19: invokeinterface #17, 1// InterfaceMethod org/slf4j/Logger.isWarnEnabled:()Z
    24: ifeq            44   
    27: getstatic       #1   // Field log:Lorg/slf4j/Logger;
    30: ldc             #18  // String Detected a null value, null values are not allowed by Conjure
    32: new             #19  // class java/lang/IllegalArgumentException
    35: dup             
    36: invokespecial   #20  // Method java/lang/IllegalArgumentException."<init>":()V
    39: invokeinterface #21, 3// InterfaceMethod org/slf4j/Logger.warn:(Ljava/lang/String;Ljava/lang/Throwable;)V
    44: return          
```

Reply:
This is wonderful, thanks! Opened a PR with the changes I made to make the build succeed.